packages/web3/src/contract/script-simulator.ts

```diff
@@ -0,0 +1,837 @@
+          } else {
+            localVariables.set(Number(index.value), operandStack.pop())
+          }
+          break
+        }
+        case 'CallerAddress': {
+          operandStack.push(callerAddress)
+          break
+        }
+        case 'ApproveAlph': {
+          const amount = operandStack.popU256() // amount
+          const spender = operandStack.popAddress() // spender
+          if (spender.kind.startsWith('Symbol')) {
+            approved.setUnknown() // The spender might be the caller
+          } else if (spender === callerAddress) {
+            approved.addApprovedAttoAlph(amount)
+          }
+          break
+        }
+        case 'ApproveToken': {
+          const amount = operandStack.popU256() // amount
+          const tokenId = operandStack.popByteVec() // token
+          const spender = operandStack.popAddress() // spender
+          if (spender.kind.startsWith('Symbol')) {
+            approved.setUnknown() // The spender might be the caller
+          } else if (spender === callerAddress) {
+            approved.addApprovedToken(tokenId, amount)
+          }
+          break
+        }
+        case 'CreateContractAndTransferToken': {
+          operandStack.popAddress() // token owner
+        }
+        case 'CreateContractWithToken': {
+          operandStack.popU256() // token amount
+        }
+        case 'CreateContract': {
+          operandStack.popByteVec() // mutable fields
+          operandStack.popByteVec() // immutable fields
+          operandStack.popByteVec() // contract code
+          operandStack.push({ kind: 'Symbol-ByteVec', value: undefined }) // new contract id
+          break
+        }
+        case 'TransferAlph': {
+          operandStack.popU256() // amount
+          operandStack.popAddress() // recipient
+          operandStack.popAddress() // sender
+          break
+        }
+        case 'TransferToken': {
+          operandStack.popU256() // amount
+          operandStack.popByteVec() // token
+          operandStack.popAddress() // recipient
+          operandStack.popAddress() // sender
+          break
+        }
+        default:
+          unimplemented(instr.name)
+          break
+      }
+    }
+    return contractCalls
+  }
+}
+
+type SymbolBool = { kind: 'Symbol-Bool'; value: undefined }
+type SymbolI256 = { kind: 'Symbol-I256'; value: undefined }
+type SymbolU256 = { kind: 'Symbol-U256'; value: undefined }
+type SymbolByteVec = { kind: 'Symbol-ByteVec'; value: undefined }
+type SymbolAddress = { kind: 'Symbol-Address'; value: undefined }
+type SymbolAny = { kind: 'Symbol-Any'; value: undefined }
+type SimulatorVal = Val | SymbolBool | SymbolI256 | SymbolU256 | SymbolByteVec | SymbolAddress | SymbolAny
+type SimulatorVar<K extends string, V> = { kind: K; value: V } | { kind: `Symbol-${K}`; value: undefined }
+
+function unaryOp<K extends string, V>(x: SimulatorVar<K, V>, op: (x: V) => V): SimulatorVar<K, V> {
+  if (x.kind.startsWith('symbol')) {
```

Comment:
Do we need to use `Symbol` instead of `symbol` in these `*Op` functions? It seems that `'Symbol-Bool'.startsWith('symbol')` returns false.